### ISSUE TITLE
Missing native libsignal-client in v0.8.4

### ISSUE THREAD

**Issue by exquo** [AsamK/signal-cli#643]
Just installed the latest signal-cli release v0.8.4 and got 

	$ signal-cli receive
	Missing required native library dependency: libsignal-client

This is on linux, and the previous signal-cli versions have worked fine without needing any manual steps from https://github.com/AsamK/signal-cli/wiki/Provide-native-lib-for-libsignal.

I've checked, and the `libsignal_jni.so` is bundled in the jar file:

	$ unzip -l signal-cli-0.8.4/lib/signal-client-java-0.8.1.jar | grep libsignal_jni.so 
		2432368  2021-06-02 18:55   libsignal_jni.so

So, not sure what's missing.

**Comment by AsamK** [comment]
That's strange, which distribution and cpu architecture do you have?

**Comment by zxyrepf** [comment]
I get the same error on Debian 10 (x86_64).

**Comment by exquo** [comment]
> Debian 10 (x86_64)

ditto :)

**Comment by exquo** [comment]
I have compiled the libsignal-client java library (turned out it requires the latest *nightly* release of the rust compiler), and replaced the original with it: 

	cd <...>/librust-client/target/releases
	zip -u <...>/signal-cli-0.8.4/lib/signal-client-java-0.8.1.jar libsignal_jni.so

Now it works fine with signal-cli v0.8.4.

For the reference, here's the [compiled binary file](https://github.com/exquo/libsignal-client-bin/blob/a38636cedf8ebbcd7cb630793e541fccccd8f539/libsignal_jni.so). 
(I guess folks shouldn't use it in their production environments, since it's some unverified binary file coming from someone on the internet; but it might be useful for troubleshooting, in e.g. a virtual machine)

Let me know if I can do anything else to troubleshoot this on my system.

**Comment by AsamK** [comment]
Maybe upstream now compiles on a newer distribution, with a more recent libc version than debian 10 has ...
I can check it later.

**Comment by jgillum** [comment]
> I get the same error on Debian 10 (x86_64).

Same here.

**Comment by AsamK** [comment]
yep, upstream libsignal-client builds the library with ubuntu latest, which has a newer glibc version.
The provided binaries now require at least glibc 2.29 and Debian 10 has glibc 2.28 ...

```
$ objdump -T libsignal_jni.so | grep GLIBC
0000000000000000      DF *UND*  0000000000000000  GLIBC_2.29  log
```

**Comment by exquo** [comment]
Another case of this (CentOS 7): https://github.com/AsamK/signal-cli/issues/606#issuecomment-831413627

**Comment by DutchForeigner** [comment]
I've got a similar error when sending a message using signal-cli v0.8.4.1:
signal-cli -u +1234567890 send -m "This is a test message" +1234567890
OpenJDK Server VM warning: You have loaded library /tmp/resource2972181199637573165.so which might have disabled stack guard. The VM will try to fix the stack guard now.
It's highly recommended that you fix the library with 'execstack -c <libfile>', or link it with '-z noexecstack'.
WARN App - WARNING: Support for new group V2 is disabled, because the required native library dependency is missing: libzkgroup
Missing required native library dependency: libsignal-client

This is on a Raspberry Pi (Debian 10 -buster based), with libsignal-client and libzkgroup updated with local builds.
I did get signal-cli v0.8.3 compiled and working, but cannot get v0.8.4.1 to work

**Comment by exquo** [comment]
@DutchForeigner

This is probably a different issue (i.e. _not_ caused by an older version of `glibc`). If you are compiling the libs yourself, the resulting `.so` files will use whichever version of `glibc` that is on your system.

Are you using the newer versions of the native libraries? (Updated since signal-cli 0.8.3). The current `libsignal-client` version is *`v0.8.1`* (not to be confused with signal-cli's own version, which is numerically not far off currently).

**Comment by bublath** [comment]
Make sure to remove the libraries and add the new ones without path.
In case somebody is interested, I uploaded a current set (tested with signal-cli 0.8.4) for Raspberry Pi (Buster) here:
https://github.com/bublath/FHEM-Signalbot/tree/main/armv7l-0.8.4

**Comment by DutchForeigner** [comment]
@exquo Hi, Thank you for that answer. That makes now sense to me. 
And yes I did recompile both of the libraries with the new version(s) (libzkgroup & libsignal-client). 

Now I think about it, and reading the 'remove' bit., it might be that I did not remove the previous version correctly

I'll try the ones posted by @bublath and post an update when I got that tested on my side.

**Comment by exquo** [comment]
FYI: I've started a [GitHub CI/CD](https://github.com/exquo/signal-libs-build) that compiles the native libraries automatically when new signal-cli versions are released. The linux version is compiled on Ubuntu 18, so should work on "older" `glibc` linuxes, including Debian 10.

**Comment by Wikinaut** [comment]
debian 10
```
WARN  ServiceConfig - Failed to call libsignal-client: /tmp/resource10147335520034765112.so: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/resource10147335520034765112.so)
Missing required native library dependency: libsignal-client
```

What's the suggested solution for it?

**Comment by bublath** [comment]
I have a precompiled library that should work with debian 10 here:
https://github.com/bublath/FHEM-Signalbot/tree/main/amd64-glibc2.27-0.10.4
Replace it with
zip -u signal-client-java-*.jar libsignal_jni.so
and it should work. Note that this is for 0.10.4 - I cannot guarantee it will also work with 0.10.4.1 as libsignal probably got updated.

**Comment by Wikinaut** [comment]
In the meantime I decided to upgrade my server from debian10 to 11 (bullseye), which solved that "problem".

**Comment by dogukanakkaya** [comment]
I am facing the same problem on Alpine Linux in Docker Container
My Dockerfile:
```
FROM alpine

USER root

RUN apk add wget zip git make rust cargo openjdk17

ARG SIGNAL_CLI_VERSION="0.10.5"
ARG LIBSIGNAL_VERSION="0.16.0"

RUN wget https://github.com/AsamK/signal-cli/releases/download/v"${SIGNAL_CLI_VERSION}"/signal-cli-"${SIGNAL_CLI_VERSION}"-Linux.tar.gz
RUN tar xf signal-cli-"${SIGNAL_CLI_VERSION}"-Linux.tar.gz -C /opt
RUN ln -sf /opt/signal-cli-"${SIGNAL_CLI_VERSION}"/bin/signal-cli /usr/local/bin/

RUN wget https://github.com/signalapp/libsignal/releases/download/v"${LIBSIGNAL_VERSION}"/libsignal_jni.so
RUN zip -d opt/signal-cli-"${SIGNAL_CLI_VERSION}"/lib/libsignal-client-*.jar libsignal_jni.so
RUN zip -u opt/signal-cli-"${SIGNAL_CLI_VERSION}"/lib/libsignal-client-*.jar libsignal_jni.so


RUN rm signal-cli-"${SIGNAL_CLI_VERSION}"-Linux.tar.gz
RUN rm -rf libsignal_jni.so
```

**Comment by octopode-us** [comment]
Hello guys, I have the same problem on windows 10 x64 here :

WARN  ServiceConfig - Failed to call libsignal-client: D:\Users\arthu\AppData\Local\resource14946054837028231560.so: Can't load this .dll (machine code=0x0) on a AMD 64-bit platform

To be clear I am a absolute rookie on those stuff, and I'm already very surprised I got that far, if someone has an idea, it would be awesome.

thanks a lot

**Comment by cuu508** [comment]
Looks like starting from signal-cli 0.12.0 and libsignal-client 0.30.0 Ubuntu 20.04 is now too old also:

```
WARN  Manager - Failed to call libsignal-client: /tmp/12585162856697034629libsignal_jni.so: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /tmp/12585162856697034629libsignal_jni.so)
Missing required native library dependency: libsignal-client
```

**Comment by mitar** [comment]
I approached this differently. I used Docker to run local Ubuntu container with the latest glibc and mount my home directory into the container. Then I installed `apt-get install openjdk-21-jre` and `adduser mitar` with same uid 1000 as I have on the host. Then I made `su mitar` to get in there with same uid as on the host. And I was able to run `signal-cli`.

```
docker run --rm -t -i -v /home/mitar:/home/mitar ubuntu:latest
```
